synthesis.go

```diff
@@ -0,0 +1,262 @@
+an in-memory SQLite database with everything about our sequence - including its codon table,
+possible codon-to-codont transformations, and the sequence itself.
+
+We then run all the user-provided problematicSequenceFuncs in-parallel to generate DnaSuggestions
+of places that we should fix in our sequence. Then, we check if there are any DnaSuggestions that
+overlap - for example, perhaps a repeat region needs to be fixed in a location with a high GC content,
+so we bias fixing that repeat region to fixes that have a high AT content. If there are any overlapping
+suggestions, those suggestions are fixed first.
+
+Once we have a list of those suggestions, we fix the sequence using some simple SQL. Once our first round
+of fixes is applied, we test the sequence again to see if we missed any problematic sequence locations. We
+do that for 1000 rounds, at which point we just give up and return the sequence to the user as is.
+
+Hopefully, this will enable more high throughput synthesis of genes!
+
+Keoni
+
+[started Dec 9, 2020]
+******************************************************************************/
+
+// DnaSuggestion is a suggestion of a fixer, generated by a problematicSequenceFunc.
+type DnaSuggestion struct {
+	Start          int    `db:"start"`
+	End            int    `db:"end"`
+	Bias           string `db:"gcbias"`
+	QuantityFixes  int    `db:"quantityfixes"`
+	SuggestionType string `db:"suggestiontype"`
+	Step           int    `db:"step"`
+	Id             int    `db:"id"`
+}
+
+// FindBsaI is a simple problematicSequenceFunc, for use in testing
+func FindBsaI(sequence string, c chan DnaSuggestion, wg *sync.WaitGroup) {
+	re := regexp.MustCompile(`GGTCTC`)
+	locs := re.FindAllStringIndex(sequence, -1)
+	for _, loc := range locs {
+		position := loc[0] / 3
+		leftover := loc[0] % 3
+		switch {
+		case leftover == 0:
+			c <- DnaSuggestion{position, (loc[1] / 3), "NA", 1, "BsaI removal", 0, 0}
+		case leftover != 0:
+			c <- DnaSuggestion{position, (loc[1] / 3) - 1, "NA", 1, "BsaI removal", 0, 0}
+		}
+	}
+	wg.Done()
+}
+
+// FindTypeIIS is a problematicSequenceFunc used for finding TypeIIS restriction enzymes. It finds BbsI, BsaI, BtgZI, BsmBI, SapI, and PaqCI(AarI)
+func FindTypeIIS(sequence string, c chan DnaSuggestion, wg *sync.WaitGroup) {
+	enzymeSites := []string{"GAAGAC", "GGTCTC", "GCGATG", "CGTCTC", "GCTCTTC", "CACCTGC"}
+	var enzymes []string
+	for _, enzyme := range enzymeSites {
+		enzymes = []string{enzyme, ReverseComplement(enzyme)}
+		for _, site := range enzymes {
+			re := regexp.MustCompile(site)
+			locs := re.FindAllStringIndex(sequence, -1)
+			for _, loc := range locs {
+				position := loc[0] / 3
+				leftover := loc[0] % 3
+				switch {
+				case leftover == 0:
+					c <- DnaSuggestion{position, (loc[1] / 3), "NA", 1, "TypeIIS removal", 0, 0}
+				case leftover != 0:
+					c <- DnaSuggestion{position, (loc[1] / 3) - 1, "NA", 1, "TypeIIS removal", 0, 0}
+				}
+			}
+		}
```

Comment:
This is the same code as `FindBsaI`.

 Make a generic `FindEnzme` function that takes it's regex and `SuggestionType` as parameters. 
 
 Then `FindBsaI` becomes:
 
 `FindEnzme(sequence string, c chan DnaSuggestion, wg *sync.WaitGroup, `GGTCTC`, "BsaI removal")`
 
 Write a unit test for this generic `FindEnzyme` function.